sky/optimizer.py

```diff
@@ -823,21 +823,42 @@ def _cloud_in_list(cloud: clouds.Cloud, lst: List[clouds.Cloud]) -> bool:
     return any(cloud.is_same_cloud(c) for c in lst)
 
 
-def _generate_launchables_with_region_zones(
-        resources: resources_lib.Resources) -> List[resources_lib.Resources]:
-    assert resources.is_launchable()
+def _make_launchables_for_valid_region_zones(
+    launchable_resources: resources_lib.Resources
+) -> List[resources_lib.Resources]:
+    assert launchable_resources.is_launchable()
+    # In principle, all provisioning requests should be made at the granularity
+    # of a single zone. However, for on-demand instances, we batch the requests
+    # to the zones in the same region in order to leverage the region-level
+    # provisioning APIs of AWS and Azure. This way, we can reduce the number of
+    # API calls, and thus the overall failover time. Note that this optimization
+    # does not affect the user cost since the clouds charge the same prices for
+    # on-demand instances in the same region regardless of the zones. On the
+    # other hand, for spot instances, we do not batch the requests because the
+    # "AWS" spot prices may vary across zones.
+
+    # NOTE(woosuk): GCP does not support region-level provisioning APIs. Thus,
+    # while we return per-region resources here, the provisioner will still
+    # issue the request for one zone at a time.
+    # NOTE(woosuk): If we support Azure spot instances, we should batch the
+    # requests since Azure spot prices are region-level.
+    # TODO(woosuk): Batch the per-zone AWS spot instance requests if they are
+    # in the same region and have the same price.
+    # FIXME(woosuk): Batching should be done at the higher level, not here.
```

Comment:
Curious: can you say more?

Reply:
It was my response to https://github.com/skypilot-org/skypilot/pull/975#discussion_r1032756799

> Btw, if we have batching implemented, then we do not need to special-case here, right? Basically on-demand requests can also have a list of (r, z1), (r, z2), ... and then batching will detect that they all have the same prices. Worth adding a comment?

Reply:
Changed to
```python
# FIXME(woosuk): Batching should be done at a higher level (e.g., by
# provisioner or optimizer), not here.
```